### ISSUE TITLE
关于对已初始化好的模型进行并发抽取图像向量问题

### ISSUE THREAD

**Issue by TheRainWalkStreet** [OFA-Sys/Chinese-CLIP#376]
您好！我将您开源的能力继承到了我的 FastAPI 项目中，在测试过程中遇到了一些我无法准确定位到原因的问题，特来请教！

我基于对 demo 案例的理解，尝试对已经初始化好的模型，将模型抽取图片向量部分逻辑封装成一个公共方法。外层 FastAPI 接口方法调用该公共方法，对传入的图片进向量抽取。在端到端交互过程中，客户端调用抽取图片向量接口是并行调用的，但是在图片向量抽取这里变成了串行处理，当遇到这个问题时我尝试将图片抽取向量部分逻辑增加多线程处理，但最终测试下来并没有得到改善。多次尝试后怀疑可能与模型的机制有关系，我分享一下关键部分代码，期望得到大佬的指导！thanks！

PS：公用提取图片向量方法 process_images_with_cn_clip

`class CnClip:
    def __init__(self):
        self.data = []

    def load_model_cache(self, operation_mode, clip_name, model_type):
        model, preprocess = load_cn_clip_model(operation_mode, clip_name)
        self.insert_model_cache(model_type, clip_name, model, preprocess)

    def insert_model_cache(self, model_type, clip_name, model, preprocess):
        if model is not None:
            self.data.append({
                'model_type': model_type,
                'clip_name': clip_name,
                'model': model,
                'preprocess': preprocess
            })
            logger.debug(f"Load {model_type} image model by {clip_name} success!")
        else:
            logger.error(f"Load {model_type} image model by '{clip_name}' failed!")

    def init_model_env(self):
        futures = []
        with concurrent.futures.ThreadPoolExecutor() as executor:
            for item in config.MODELS:
                clip_name = item['clip_name']
                operation_mode = item['operation_mode']
                cn_clip_is_enabled = item['is_enable']
                if cn_clip_is_enabled:
                    if clip_name in _MODELS:
                        _download(_MODELS[clip_name], "checkpoints")
                        futures.append(
                            executor.submit(self.load_model_cache, operation_mode, clip_name, 'process_images'))
                        futures.append(
                            executor.submit(self.load_model_cache, operation_mode, clip_name, 'image_search'))
                    else:
                        logger.debug(f"Invalid model [{clip_name}]!")
                else:
                    logger.debug(f"Cn clip [{clip_name}] not enabled")

            for future in concurrent.futures.as_completed(futures):
                try:
                    future.result()
                except Exception as e:
                    logger.error(f"Model loading failed: {e}")
        size_in_bytes = asizeof.asizeof(self.data)
        size_in_mb = size_in_bytes / (1024 * 1024)
        logger.debug(f"The memory size occupied by the initialized model is {size_in_mb:.6f} MB")

    def get_cn_clip_model_and_preprocess(self, model_type, clip_name):
        model, preprocess = None, None
        for item in self.data:
            if item['model_type'] == model_type and item['clip_name'] == clip_name:
                model = item['model']
                preprocess = item['preprocess']
        return model, preprocess

    async def process_images_with_cn_clip(self, clip_name, image_info_list: list[ImageData] = None):
        def process_single_image(item, model, preprocess, device, clip_dict_queue):
            try:
                image_url = item.url
                image_base64_str = item.base64_str
                img = None
                if is_not_empty(image_base64_str):
                    img = load_base64_image(image_base64_str)
                if img is None and is_not_empty(image_url):
                    img = download_image_from_url(image_url)
                if img is None:
                    return
                img = preprocess(img).unsqueeze(0).to(device)
                with torch.no_grad():
                    image_features = model.encode_image(img)
                    image_features /= image_features.norm(dim=-1, keepdim=True)
                object_key = get_image_data_key(item)
                clip_dict_queue.put((object_key, image_features.cpu()))
            except Exception as e:
                logger.debug(f"Error processing item {item}: {e}")

        def get_image_data_key(item: ImageData):
            key_json_object = {
              # ...
            }
            return json.dumps(key_json_object, separators=(',', ':'))

        model, preprocess = self.get_cn_clip_model_and_preprocess('process_images', clip_name)
        device = get_cn_clip_operation_mode(clip_name)
        if device is None or model is None:
            return None
        clip_dict_queue = queue.Queue()
        tasks = [
            asyncio.to_thread(process_single_image, item, model, preprocess, device, clip_dict_queue)
            for item in image_info_list
        ]
        await asyncio.gather(*tasks)
        clip_dict = {}
        while not clip_dict_queue.empty():
            key, value = clip_dict_queue.get()
            clip_dict[key] = value

        return clip_dict

    def get_text_features(self, clip_name, text):
        model, _ = self.get_cn_clip_model_and_preprocess('image_search', clip_name)
        device = get_cn_clip_operation_mode(clip_name)
        if model is not None and text is not None:
            text_tokens = clip.tokenize([text]).to(device)
            with torch.no_grad():
                text_features = model.encode_text(text_tokens)
            text_features /= text_features.norm(dim=-1, keepdim=True)
            return text_features.tolist()[0]
        else:
            return None`
